### DISCUSSION TITLE
cc1plus: error: to generate dependencies you must specify either -M or -MM

### DISCUSSION THREAD

**Discussion by connectrajeev** [mesonbuild/meson#13669] [General]
Hi All, 

  I am trying to build static library by invoking static_library() within a docker container which has x86 based 64-bit toolchain install in it.
  I am able to setu p the builddir however when I issue the 'meson compile' I get the following errro for each file mentioned in the sources. I am not able to fix it as its meson which is passing -MD,  -MQ &  -MF flag.

  "cc1plus: error: to generate dependencies you must specify either -M or -MM"
  This is how my function looks like:
```meson
common_cflags = [
    '-O2',
    '-W',
    '-Wall',
    '-Werror',
    '-Wno-address',
    '-Wno-cast-function-type',
    '-Wno-class-memaccess',
    '-Wno-error=type-limits',
    '-Wno-format-truncation',
    '-Wno-ignored-qualifiers',
    '-Wno-implicit-fallthrough',
    '-Wno-logical-not-parentheses',
    '-Wno-misleading-indentation',
    '-Wno-narrowing',
    '-Wno-nonnull-compare',
    '-Wno-parentheses',
    '-Wno-stringop-overflow',
    '-Wno-stringop-truncation',
    '-Wno-unknown-pragmas',
    '-Wno-unused-function',
    '-Wno-unused-parameter',
    '-fno-omit-frame-pointer',
    '-fno-strict-aliasing',
    '-ftrack-macro-expansion=0',
    '-m64',
    '-march=core2',
    '-mtune=core2',
    '-nostdinc',
    '-isystem'
    ]

   static_library (name,
            sources,
            include_directories: includes,
            cpp_args: common_cflags,
            build_by_default: true,
            install: true,
            install_dir: csd + name + '/' + common_install
        )
```

**Comment by dcbaker** [comment]
What compiler are you using, and what version of that compiler?

**Reply by connectrajeev** [reply]
bash-4.4$ /usr/bin/x86_64-redhat-linux-g++ --version
x86_64-redhat-linux-g++ (GCC) 8.5.0 20210514 (Red Hat 8.5.0-22.0.1)
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
on oracle-linux-8.

**Reply by connectrajeev** [reply]
Any suggestions on how to approach this issue? I'm not expecting an exact solution, but I would appreciate guidance on the areas I should investigate, as I'm currently unsure of how to address it.

**Reply by dcbaker** [reply]
I'm pretty baffled, I know that very old versions of GCC don't support the flags we use, but that's 3.x and 4.x.

I've seen some bug reports that look really similar that turned out to be ccache? If you have ccache installed can you try `CXX=/usr/bin/x86_64-redhat-linux-g++ meson setup $builddir? That will prevent Meson from detecting ccache

**Reply by connectrajeev** [reply]
Thanks @dcbaker. Well, I don't have ccache installed in my container but I was planning to install it if everything goes well without ccache. 
My native.txt file looks something like this:
[binaries]
distcc = 'distcc'
dmucs = 'dmucs_gethost'
c   = '/usr/bin/x86_64-redhat-linux-gcc'
cpp = '/usr/bin/x86_64-redhat-linux-g++'

and in meson.build, I am setting up the compiler like this and I am not using the distcc and dmucs as well. 
cc =  meson.get_compiler('c')
cpp = meson.get_compiler('cpp')

Is there a way, I can instruct meson not to add the default options from its side?

**Reply by dcbaker** [reply]
No, we always attempt to write out a depfile because re-compilation isn't reliable without one unless you manually specify every header each C++ file uses 100% correctly for every platform.

What's really confusing is that you should have in the `C_COMPILER` rule of your build.ninja something like `command = gcc $ARGS -MD -MQ $out -MF $DEPFILE -o $out -c $in`. `-MD` is a `-M` rule, but creates the depfile as a side effect and still does regular copilation...

can you try running `/usr/bin/x86_64-redhate-linux-gcc -MD -MQ obj.o -MF obj.d -o obj.o -c obj.c` with an super basic .c file? Something like `int foo(void) { return 0; }` is fine.

**Reply by connectrajeev** [reply]
Thank you @dcbaker ! I could resolve the issue - there was a problem in the `common_cflags` I have defined. I was passing both `-nostdinc` & `-isystem` looks like gcc is not liking that but I am not sure why it is so. For now, I have removed both flags from `common_cflags`. 
Do you know why we can't pass both `-nostdinc` & `-isystem`? 
Thanks again for your help!
